python/revng/model/metaaddress.py

```diff
@@ -0,0 +1,162 @@
+            self.__reject_additional_kwargs(kwargs)
+
+            addr = self.__get_arg(args, kwargs, 0, "Address")
+            self.Address = self.__parse_int(addr)
+
+            _type = self.__get_arg(args, kwargs, 1, "Type")
+            self.Type = MetaAddressType(_type)
+
+            epoch = self.__get_arg(args, kwargs, 2, "Epoch", mandatory=False)
+            if epoch is not None:
+                self.Epoch = self.__parse_int(epoch)
+
+            addr_space = self.__get_arg(args, kwargs, 3, "AddressSpace", mandatory=False)
+            if addr_space is not None:
+                self.AddressSpace = self.__parse_int(addr_space)
+
+        else:
+            raise ValueError("Wrong arguments to MetaAddress constructor. MetaAddress can be constructed by:\n"
+                             " - providing a serialized MetaAddress as a string, or\n"
+                             " - providing Address and Type and optionally Epoch and AddressSpace")
+
+    def _parse_string(self, s: str):
+        assert isinstance(s, str)
+
+        match = self.yaml_regexp.fullmatch(s)
+        if match is None:
+            raise ValueError(f"Could not parse {s} as a MetaAddress")
+
+        address = match["Address"] or "0"
+        meta_address_type = match["Type"]
+        epoch = match["Epoch"] or "0"
+        address_space = match["AddressSpace"] or "0"
+
+        self.Address = int(address, base=0)
+        self.Type = MetaAddressType[meta_address_type]
+        self.Epoch = int(epoch, base=0)
+        self.AddressSpace = int(address_space, base=0)
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        return cls(**kwargs)
+
+    @classmethod
+    def from_string(cls, s: str):
+        return cls(s)
+
+    def is_default_epoch(self):
+        return self.Epoch == 0
+
+    def is_default_address_space(self):
+        return self.AddressSpace == 0
+
+    def is_invalid(self):
+        return self.Type == MetaAddressType.Invalid
+
+    def __repr__(self):
+        if self.Address == 0:
+            addr = ""
+        else:
+            addr = hex(self.Address)
+
+        components = [
+            addr,
+            self.Type.name,
+        ]
+        if not self.is_default_epoch():
+            components.append(str(self.Epoch))
+        if not self.is_default_address_space():
+            components.append(str(self.AddressSpace))
+        return ":".join(components)
+
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: "MetaAddress"):
+        return dumper.represent_str(repr(instance))
+
+    @staticmethod
+    def __reject_additional_kwargs(kwargs: dict):
+        additional_kwargs = kwargs.keys() - {"Address", "Type", "Epoch", "AddressSpace"}
+        if additional_kwargs:
+            raise ValueError(f"Unexpected arguments supplied to MetaAddress constructor: {additional_kwargs}")
+
+    @staticmethod
+    def __get_arg(args, kwargs, idx, name, mandatory=True, default_if_not_mandatory=None):
```

Comment:
* Why `__`?
* This all seems quite complicated. What's the goal? Is it strictly necessary?

Reply:
`__` because AFAIK it ensures that I'm not overriding something from the dataclasses base class.
Re: 2nd question, it's a helper to get an argument that can be passed either as positional or as keyword, see the thread above